### DISCUSSION TITLE
useSWRSubscription: how to preserve type in next() callback?

### DISCUSSION THREAD

**Discussion by xaphod** [vercel/swr#2522] [Q&A]
```
const _useDocListener = <T extends DocumentData>(
  app: FirebaseApp,
  collectionPath?: string | null,
  id?: string | null,
) => {
  const subscribe: SWRSubscription<WithIDAndRef<T>, FirestoreError> = (key, { next }) => {
    const [path, id] = key as Array<string | null | undefined>
    if (!path || !id) return () => null

    return onSnapshot(
      doc(getFirestore(app), path, id),
      (snapshot) => {
        const newData = { ...snapshot.data(), id: snapshot.id, ref: snapshot.ref } as WithIDAndRef<T>
        return next(null, newData)
      },
      (e: FirestoreError) => {
        console.error('useEvent onSnapshot error:')
        console.error(e)
        return next(e)
      },
    )
  }

  const { data } = useSWRSubscription([collectionPath, id], subscribe)
}
```

Expected: `data` is type `WithIDAndRef<T>`
Actual: `data` is type `any`

Other things I tried:
- don't use `SWRSubscription` type, just let type inference do its thing... no difference

All help appreciated!

**Comment by xaphod** [comment]
... so far my workaround is

```
  const typedUseSWRSubscription = useSWRSubscription as SWRSubscriptionHook<WithIDAndRef<T>, FirestoreError>
```

which feels kinda gross

**Reply by huozhi** [reply]
Does using generics work for you?

```ts
const { data } = useSWRSubscription<WithIDAndRef<T>, FirestoreError>(...)
```

it's indeed sth could improve on the auto inferring of the types

**Reply by xaphod** [reply]
(fix screenshot)
No, you can't specify generics to `useSWRSubscription`:

<img width="1246" alt="Screen Shot 2023-03-22 at 5 06 06 PM" src="https://user-images.githubusercontent.com/3776302/227038140-26173bb4-03cb-4d80-b635-94030fa627d6.png">

**Reply by huozhi** [reply]
Ah my bad, it's not available, will look into the auto inferring for the return value

**Comment by huozhi** [comment]
https://swr.vercel.app/docs/typescript#useswrsubscription

We've improved the types inferring in [2.1.2](https://github.com/vercel/swr/releases/tag/v2.1.2), please give it another try on that! 🙏 thanks!

**Reply by xaphod** [reply]
Hmm that didn't change anything for me - when i pass in strongly typed parameters to the callback, the return value from the hook is still `any`:

<img width="1026" alt="Screenshot 2023-04-05 at 8 42 58 AM" src="https://user-images.githubusercontent.com/3776302/230083714-413db107-38de-40b6-94ba-14d67b3aad9b.png">
